prov/sockets/src/sock_rx_entry.c

```diff
@@ -49,30 +49,54 @@
 #define SOCK_LOG_DBG(...) _SOCK_LOG_DBG(FI_LOG_EP_DATA, __VA_ARGS__)
 #define SOCK_LOG_ERROR(...) _SOCK_LOG_ERROR(FI_LOG_EP_DATA, __VA_ARGS__)
 
-/* FIXME: pool of rx_entry */
 struct sock_rx_entry *sock_rx_new_entry(struct sock_rx_ctx *rx_ctx)
 {
 	struct sock_rx_entry *rx_entry;
+	struct slist_entry *entry;
+	int i;
 
-	rx_entry = calloc(1, sizeof(*rx_entry));
-	if (!rx_entry)
-		return NULL;
+	if (rx_ctx->rx_entry_pool == NULL) {
+		rx_ctx->rx_entry_pool = calloc(rx_ctx->attr.size, sizeof (*rx_entry));
+		slist_init(&rx_ctx->pool_list);
+		
+		for (i = 0; i < rx_ctx->attr.size; i++) {
+			slist_insert_tail(&rx_ctx->rx_entry_pool[i].pool_entry,
+					  &rx_ctx->pool_list);
+			rx_ctx->rx_entry_pool[i].is_pool_entry = 1;
+		}
+	}
+
+	if (!slist_empty(&rx_ctx->pool_list)) {
+		entry = slist_remove_head(&rx_ctx->pool_list);
+		rx_entry = container_of(entry, struct sock_rx_entry, pool_entry);
+		rx_entry->rx_ctx = rx_ctx;
+		entry = slist_remove_head(&rx_ctx->pool_list);
+	} else {
+		rx_entry = calloc(1, sizeof(*rx_entry));
+		if (!rx_entry)
+			return NULL;
```

Comment:
Ideally, application should not post more than the rx attr size. But I thought of handling the case where user posts more than the rx size.